modules/apps/foundation/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_permissions_navigation.jspf

```diff
@@ -145,9 +145,9 @@ PanelCategoryRegistry panelCategoryRegistry = (PanelCategoryRegistry)request.get
 											for (PanelApp panelApp : panelApps) {
 												Portlet panelAppPortlet = PortletLocalServiceUtil.getPortletById(company.getCompanyId(), panelApp.getPortletId());
 
-												if (Objects.equals(panelAppPortlet.getControlPanelEntryClass(), com.liferay.portal.kernel.portlet.OmniadminControlPanelEntry.class.getName()) ||
-														com.liferay.portal.kernel.portlet.OmniadminControlPanelEntry.class.isAssignableFrom(panelAppPortlet.getControlPanelEntryInstance().getClass())) {
+												ControlPanelEntry controlPanelEntry = panelAppPortlet.getControlPanelEntryInstance();
 
+												if (Objects.equals(panelAppPortlet.getControlPanelEntryClass(), OmniadminControlPanelEntry.class.getName()) || OmniadminControlPanelEntry.class.isAssignableFrom(controlPanelEntry.getClass())) {
```

Comment:
@hhuijser @ling-alan-huang  this check seems redundant to me, no string comparison should be done. panelAppPortlet.getControlPanelEntryInstance() shall never be null, even if we create a portlet using the XML descriptor. PortletImpl relies on PortletBag to ask for the control panel entry instance. PortletBag in turn gets that info from the PortletBagFactory. We use the factory to create the portlet bag both from the PortletTracker and the PortletLocalServiceImpl.initWar so we should expect that both mechanism (xml and component) will give us the same result in terms of  getControlPanelEntryInstance(). 

The definite test shall be to deploy an xml-based portlet declaring the omniadmin control panel class and test if it gets hidden from the menu.

@ling-alan-huang  could you please test this?
Thank you

Reply:
Hi @dsanz,
    cc @hhuijser,

Sorry for my late reply.

I removed the Object comparison, but "Meetings Admin" will show in the page.
"Meetings Admin" is hidden before we fix this bug, and I think this portlet is created by
using XML descriptor.

Here is my test output for "LicenseManager" and "Meetings Admin".
,,,,,,,,,,,,,,,,,,,,,,
panelApp.getPortletId():com_liferay_license_manager_web_portlet_LicenseManagerPortlet
panelAppPortlet.getPortletName():com_liferay_license_manager_web_portlet_LicenseManagerPortlet
panelAppPortlet.getPortletClass():com.liferay.license.manager.web.internal.portlet.LicenseManagerPortlet
panelAppPortlet.getControlPanelEntryClass():null
panelAppPortlet.getControlPanelEntryInstance().getClass():class com.liferay.license.manager.web.internal.portlet.LicenseManagerControlPanelEntry
,,,,,,,,,,,,,,,,,,,,,,
panelApp.getPortletId():2_WAR_powwowportlet
panelAppPortlet.getPortletName():2
panelAppPortlet.getPortletClass():com.liferay.powwow.admin.portlet.AdminPortlet
panelAppPortlet.getControlPanelEntryClass():com.liferay.portal.kernel.portlet.OmniadminControlPanelEntry
panelAppPortlet.getControlPanelEntryInstance().getClass():class com.sun.proxy.$Proxy853

Reply:
Thanks for the tests @ling-alan-huang.

So it seems that we can obtain a control panel entry instance from portlets having a XML descriptor, but that instance seems not directly assignable to the omniadmin control panel entry because it's a proxy. I'm not sure we can safely obtain the proxied class. In case of doubt we'll keep both mechanisms. Let me check this, will come back to you later

Best

Reply:
Hi @ling-alan-huang I believe it's better to keep both mechanisms as I see this lookup as a potentially risky operation.